data/src/test/java/org/apache/iceberg/io/TestWriterMetrics.java

```diff
@@ -0,0 +1,150 @@
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+
+import static org.apache.iceberg.types.Types.NestedField.optional;
+import static org.apache.iceberg.types.Types.NestedField.required;
+
+@RunWith(Parameterized.class)
+public abstract class TestWriterMetrics<T> {
+
+  private static final int FORMAT_V2 = 2;
+
+  protected static final Types.NestedField ID_FIELD = required(1, "id", Types.IntegerType.get());
+  protected static final Types.NestedField DATA_FIELD = optional(2, "data", Types.StringType.get());
+
+  protected static final Types.StructType NESTED_FIELDS = Types.StructType.of(
+      required(4, "booleanField", Types.BooleanType.get()),
+      optional(5, "longValue", Types.LongType.get()));
+
+  protected static final Types.NestedField STRUCT_FIELD = optional(3, "structField", NESTED_FIELDS);
+
+  // create a schema with all supported fields
+  protected static final Schema SCHEMA = new Schema(
+      ID_FIELD,
+      DATA_FIELD,
+      STRUCT_FIELD
+  );
+
+  protected static final SortOrder sortOrder =
+      SortOrder.builderFor(SCHEMA).asc("id").asc("structField.longValue").build();
+
+  protected static final Map<String, String> properties =
+      ImmutableMap.of(TableProperties.DEFAULT_WRITE_METRICS_MODE, "none");
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  protected FileFormat fileFormat;
+  protected TestTables.TestTable table = null;
+  protected File metadataDir = null;
+  private File tableDir = null;
+  private OutputFileFactory fileFactory = null;
+
+  @Parameterized.Parameters(name = "FileFormat = {0}")
+  public static Object[][] parameters() {
+    return new Object[][] {
+        {FileFormat.ORC},
+        {FileFormat.PARQUET}
+    };
+  }
+
+  public TestWriterMetrics(FileFormat fileFormat) {
+    this.fileFormat = fileFormat;
+  }
+
+  protected abstract WriterFactory newWriterFactory(Schema dataSchema);
+
+
+  protected abstract T toRow(Integer id, String data, boolean boolValue, Long longValue);
+
+  @Before
+  public void setupTable() throws Exception {
+    this.tableDir = temp.newFolder();
+    tableDir.delete(); // created by table create
+
+    this.metadataDir = new File(tableDir, "metadata");
+
+    this.table = TestTables.create(
+        tableDir,
+        "test",
+        SCHEMA,
+        PartitionSpec.unpartitioned(),
+        sortOrder,
+        FORMAT_V2);
+    table.updateProperties().set(TableProperties.DEFAULT_WRITE_METRICS_MODE, "none").commit();
+
+    this.fileFactory = OutputFileFactory.builderFor(table, 1, 1).format(fileFormat).build();
+  }
+
+  @After
+  public void after() {
+    TestTables.clearTables();
+  }
+
+  @Test
+  public void verifySortedColMetric() throws Exception {
+    T row = toRow(3, "3", true, 3L);
+    FileAppender<T> appender = newWriterFactory(SCHEMA).newDataWriter(
```

Comment:
I think you can use `DataWriter#toDataFile` to get accesses to bounds and avoid exposing the appender.

Reply:
Done